File: presentation/src/main/java/io/tripled/social/client/presentation/ReadEvalPrintLoop.java
package io.tripled.social.client.presentation;

import io.tripled.social.client.presentation.controller.FollowController;
import io.tripled.social.client.presentation.controller.PostController;
import io.tripled.social.client.presentation.controller.ReadController;
import io.tripled.social.client.presentation.controller.WallController;
import io.tripled.social.client.presentation.request.CliRequestFactory;
import io.tripled.social.client.presentation.request.RequestDispatcher;

import java.util.Objects;

public class ReadEvalPrintLoop implements Runnable {

    private static final String QUIT_COMMAND = "\\q";
    private final Input input;
    private final Output output;
    private final CliRequestFactory requestFactory;
    private final RequestDispatcher dispatcher;
    private boolean quitCommandReceived;

    public ReadEvalPrintLoop(Input input,
                             Output output,
                             PostController postController,
                             ReadController readController,
                             WallController wallController,
                             FollowController followController) {
        Objects.requireNonNull(input, "the Input cannot be null");
        Objects.requireNonNull(output, "the Output cannot be null");
        this.input = input;
        this.output = output;
        this.requestFactory = new CliRequestFactory();
        this.dispatcher = new RequestDispatcher(postController, readController, wallController, followController);
    }

    public boolean isStopped() {
        return quitCommandReceived;
    }

    @Override
    public void run() {
        output.print("Started ...");
        quitCommandReceived = false;
        while (!quitCommandReceived) {
            String readLine = input.read();
            if (isQuitCommand(readLine)) {
                quitCommandReceived = true;
                break;
            }
            executeCommands(readLine);
        }
        output.print("Finished");
    }

    private void executeCommands(String readLine) {
        output.print("read command [" + readLine + "]");
        requestFactory.create(readLine)
                      .ifPresent(request -> request.accept(dispatcher, output));
    }

    private boolean isQuitCommand(String readLine) {
        return QUIT_COMMAND.equals(readLine);
    }
}

File: src/main/kotlin/io/tripled/social/client/infrastructure/cli/ReadEvalPrintLoop.kt
package io.tripled.social.client.infrastructure.cli

import io.tripled.social.client.infrastructure.cli.controller.FollowController
import io.tripled.social.client.infrastructure.cli.controller.PostController
import io.tripled.social.client.infrastructure.cli.controller.ReadController
import io.tripled.social.client.infrastructure.cli.controller.WallController
import io.tripled.social.client.infrastructure.cli.request.CliRequestFactory
import io.tripled.social.client.infrastructure.cli.request.RequestDispatcher
import java.util.*

class ReadEvalPrintLoop(
    input: Input,
    output: Output,
    postController: PostController,
    readController: ReadController,
    wallController: WallController,
    followController: FollowController
) : Runnable {
    private val input: Input
    private val output: Output
    private val requestFactory: CliRequestFactory
    private val dispatcher: RequestDispatcher
    var isStopped: Boolean = false
        private set

    init {
        Objects.requireNonNull<Input?>(input, "the Input cannot be null")
        Objects.requireNonNull<Output?>(output, "the Output cannot be null")
        this.input = input
        this.output = output
        this.requestFactory = CliRequestFactory()
        this.dispatcher = RequestDispatcher(postController, readController, wallController, followController)
    }

    override fun run() {
        this.isStopped = false
        while (!this.isStopped) {
            input.read()?.let {
                if (isQuitCommand(it)) {
                    this.isStopped = true
                    break
                }
                executeCommands(it)
            }
        }
    }

    private fun executeCommands(readLine: String) {
        requestFactory.create(readLine)?.accept(dispatcher, output)

    }

    private fun isQuitCommand(readLine: String): Boolean {
        return QUIT_COMMAND == readLine
    }

    companion object {
        private const val QUIT_COMMAND = "\\q"
    }
}
